### PR TITLE
#9 Add support for HCL files

### PR DESCRIPTION
**Pull request by prestonvasquez**
Resolves https://github.com/Boeing/config-file-validator/issues/9

This ticket proposes using [Parser.ParseHCL ](https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclparse#Parser.ParseHCL) from the [hclparse](https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclparse) package to validate encoded .hcl files. This works by "[peeking](https://github.com/hashicorp/hcl/blob/main/hclsyntax/peeker.go#L73)" at the encoded data. The decoding methods defined by the hcl [guide](https://github.com/hashicorp/hcl/tree/main/guide) are unsuitable for validation, see hcl issue [#505](https://github.com/hashicorp/hcl/issues/505).

FWIW this solution also benchmarks slightly better than the decoder:
```
goos: darwin
goarch: arm64
pkg: github.com/Boeing/config-file-validator/pkg/validator
BenchmarkValidate-10              566096              2037 ns/op            2281 B/op         21 allocs/op
```

[hclsimple.Decode](https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclsimple#Decode):
```
goarch: arm64
pkg: github.com/Boeing/config-file-validator/pkg/validator
BenchmarkValidate-10              581282              2076 ns/op            2393 B/op         25 allocs/op
PASS
ok      github.com/Boeing/config-file-validator/pkg/validator   2.370s
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,13 +3,23 @@ module github.com/Boeing/config-file-validator
 go 1.21
 
 require (
```
Comment (prestonvasquez) [inline go.mod:5]: Changes to the go.mod file were triggered by `[go mod tidy](https://go.dev/ref/mod#go-mod-tidy)`.

File: pkg/validator/validator_test.go [hunk 2]
```diff
@@ -22,21 +22,31 @@ var testData = []struct {
...
-		}
-
-		if valid && err != nil {
-			t.Error("incorrect result: err was not nil", err)
-		}
-
-		if !valid && err == nil {
-			t.Error("incorrect result: function returned a nil error")
-		}
+	t.Parallel()
```
Comment (prestonvasquez) [inline pkg/validator/validator_test.go:30]: Parallelize tests

File: pkg/validator/validator_test.go [hunk 3]
```diff
@@ -22,21 +22,31 @@ var testData = []struct {
 	{"invalidIni", []byte(`\nCatalog hidden\n`), false, IniValidator{}},
 	{"validProperties", []byte("key=value\nkey2=${key}"), true, PropValidator{}},
 	{"invalidProperties", []byte("key=${key}"), false, PropValidator{}},
+	{"validHcl", []byte(`foo = "bar"`), true, HclValidator{}},
+	{"invalidHcl", []byte(`"foo" = "bar"`), false, HclValidator{}},
```
Comment (jackswiney) [inline pkg/validator/validator_test.go:26]: Please avoid using words like "foo" and "bar".

File: test/fixtures/subdir2/bad.hcl [hunk 4]
```diff
@@ -0,0 +1 @@
+"foo" = "bar"
```
Comment (jackswiney) [inline test/fixtures/subdir2/bad.hcl:1]: Same as above
> Please avoid using words like "foo" and "bar".

File: cmd/validator/validator.go [hunk 5]
```diff
@@ -2,7 +2,7 @@
 Validator recusively scans a directory to search for configuration files and
 validates them using the go package for each configuration type.
 
-Currently json, yaml, toml, xml and ini configuration file types are supported.
+Currently json, yaml, toml, xml, ini, and hcl configuration file types are supported.
```
Comment (kehoecj) [inline cmd/validator/validator.go:5]: Properties files are also supported - we must have missed this function doc for the last PR - please add. Thanks!

File: index.md [hunk 6]
```diff
@@ -98,7 +99,7 @@ validator --exclude-dirs=/path/to/search/tests /path/to/search
 ![Exclude Dirs Run](./img/exclude_dirs.png)
 
 #### Exclude file types
-Exclude file types in the search path. Available file types are `ini`, `json`, `yaml`, `yml`, `toml`, and `xml`
+Exclude file types in the search path. Available file types are `ini`, `json`, `yaml`, `yml`, `toml`, `xml`, and `hcl`
```
Comment (kehoecj) [inline index.md:102]: Add properties

File: index.md [hunk 7]
```diff
@@ -48,6 +48,7 @@ How many deployments have you done that needed to be rolled back due to a missin
 * YAML
 * TOML
 * INI
+* HCL
```
Comment (kehoecj) [inline index.md:51]: Add properties

File: index.md
Comment (kehoecj) [inline index.md:1]: Need to update README.md as well

**Review by kehoecj — CHANGES_REQUESTED** [review]
Looks great - some minor changes and this should be GTG

File: index.md
Comment (jd4235) [reply index.md:1]: README updates should include new *.png images showing hcl results in the screenshots.

File: pkg/validator/hcl.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+// If the parsing error does not produce an hcl.Diagnostics slice, a generic
+// error will be returned, wrapping the input file.
+func (hclv HclValidator) Validate(b []byte) (bool, error) {
+	_, diags := hclparse.NewParser().ParseHCL(b, "")
+	if diags == nil {
+		return true, nil
+	}
+
+	if len(diags) == 0 {
+		return false, fmt.Errorf("error validating .hcl file: %w", diags)
```
Comment (jd4235) [inline pkg/validator/hcl.go:33]: We need a test that covers this case as well as the case mentioned in the comment above where hcl.Diagnostics slice contains more than one error.  I only see the following in the test output.

```
    × /test2/test/fixtures/subdir2/bad.hcl
        error: error at line 1 column 1: :1,1-2: Invalid argument name; Argument names must not be quoted.
```

Comment (prestonvasquez) [reply pkg/validator/hcl.go:33]: I think it's actually best to remove this block, it should be the same as [`diags == nil`](https://go.dev/play/p/6DPkvnPiT1R) .

Comment (jd4235) [reply pkg/validator/hcl.go:33]: That's fine to remove but we still need a test that validates your notes about the diags.

File: index.md
Comment (kehoecj) [reply index.md:1]: I'd like the screenshots to look exactly the same for consistency. I think I'll go ahead and create an issue to re-do the screenshots before release v1.5.0 to grab the new validator type output. @prestonvasquez can you please revert this change? Thanks!

Comment (kehoecj) [reply index.md:1]: Created https://github.com/Boeing/config-file-validator/issues/60

**Review by kehoecj — APPROVED** [review]

File: pkg/validator/hcl.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+// If the hcl.Diagnostics slice contains more than one error, the wrapped
+// error returned by this function will include them as "and {count} other
+// diagnostic(s)" in the error message.
+func (hclv HclValidator) Validate(b []byte) (bool, error) {
+	_, diags := hclparse.NewParser().ParseHCL(b, "")
+	if diags == nil {
+		return true, nil
+	}
+
+	subject := diags[0].Subject
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply pkg/validator/hcl.go:29]: The explanation is in the comment:

> // If the hcl.Diagnostics slice contains more than one error, the wrapped
// error returned by this function will include them as "and {count} other
// diagnostic(s)" in the error message.

Here is the source code if you want to review how it works: https://github.com/hashicorp/hcl/blob/main/diagnostic.go

This case is tested, see "multipleInvalidHcl" .

[the remark this answers is not in the document]
Comment (jd4235) [reply pkg/validator/hcl.go:29]: Thanks for clarifying. I see that now.

**Comment by kehoecj** [comment]
Thank you @prestonvasquez for the contribution! 🎉

**Pull request merged**